File: r/tests/testthat/test_competition_canvas_winset.R
# test_competition_canvas_winset.R — Unit tests for animate_competition_canvas WinSet feature.

# Fixture: 3 2D voters, 2 sticker competitors, 1 seat, 2 rounds.
# Competitor 0 at (0,0) wins every round (closer to 2 of 3 voters).
# This gives n_frames = 3 (Round 1, Round 2, Final) and 1 seat holder per frame.
.ws_trace_2d <- function() {
  competition_run(
    competitor_positions = c(0.0, 0.0, 2.0, 0.0),
    strategy_kinds       = c("sticker", "sticker"),
    voter_ideals         = c(-0.5, 0.0, 0.0, 0.2, 1.8, 0.1),
    dist_config          = make_dist_config(n_dims = 2L),
    engine_config        = make_competition_engine_config(
      seat_count = 1L,
      seat_rule  = "plurality_top_k",
      max_rounds = 2L,
      step_config = make_competition_step_config(
        kind = "fixed", fixed_step_size = 0.5
      )
    )
  )
}

WS_VOTERS <- c(-0.5, 0.0, 0.0, 0.2, 1.8, 0.1)  # 3 voters, 2D flat


test_that("compute_winset = FALSE leaves winsets absent from payload", {
  skip_without_lib()
  trace <- .ws_trace_2d()
  w <- animate_competition_canvas(trace, voters = WS_VOTERS)
  expect_false("winsets" %in% names(w$x))
  expect_false("winset_competitor_indices" %in% names(w$x))
})

test_that("compute_winset = TRUE produces correctly shaped winset payload", {
  skip_without_lib()
  trace <- .ws_trace_2d()
  w <- animate_competition_canvas(
    trace,
    voters         = WS_VOTERS,
    compute_winset = TRUE
  )
  d <- trace$dims()
  n_frames <- d$n_rounds + 1L   # 3

  expect_true("overlays_frames" %in% names(w$x))
  expect_false("winsets" %in% names(w$x))
  expect_false("winset_competitor_indices" %in% names(w$x))

  frame_overlays <- w$x$overlays_frames

  expect_length(frame_overlays, n_frames)

  for (f in seq_len(n_frames)) {
    canonical_entries <- frame_overlays[[f]]$winsets
    expect_gte(length(canonical_entries), 1L)
    expect_true(all(vapply(canonical_entries, function(e) {
      e$candidate >= 0L && e$candidate < d$n_competitors
    }, logical(1L))))
  }
})

test_that("compute_winset = TRUE without voters raises an error", {
  skip_without_lib()
  trace <- .ws_trace_2d()
  expect_error(
    animate_competition_canvas(trace, compute_winset = TRUE),
    "requires voters"
  )
})

test_that("compute_winset = TRUE with winset_max exceeded raises an error", {
  skip_without_lib()
  trace <- .ws_trace_2d()
  # 1 seat × 3 frames = 3, which exceeds winset_max = 1.
  expect_error(
    animate_competition_canvas(
      trace,
      voters      = WS_VOTERS,
      compute_winset = TRUE,
      winset_max     = 1L
    ),
    "winset_max"
  )
})

test_that("non-empty winset entry has paths, is_hole, and competitor_idx keys", {
  skip_without_lib()
  trace <- .ws_trace_2d()
  w <- animate_competition_canvas(
    trace,
    voters         = WS_VOTERS,
    compute_winset = TRUE
  )
  ws <- w$x$overlays_frames
  d  <- trace$dims()

  # Find the first non-null entry.
  found <- FALSE
  for (f in seq_along(ws)) {
    for (s_i in seq_along(ws[[f]]$winsets)) {
      entry <- ws[[f]]$winsets[[s_i]]
      if (!is.null(entry)) {
        expect_true("polygons" %in% names(entry))
        expect_true("candidate" %in% names(entry))

        expect_gte(length(entry$polygons), 1L)
        first_ring <- entry$polygons[[1]]$ring
        expect_true(is.list(first_ring))
        expect_gte(length(first_ring), 2L)
        expect_true(all(vapply(first_ring, length, integer(1L)) == 2L))
        expect_true(all(is.numeric(unlist(first_ring))))

        expect_true(all(vapply(entry$polygons, function(poly) "hole" %in% names(poly), logical(1L))))

        expect_gte(entry$candidate, 0L)
        expect_lt(entry$candidate, d$n_competitors)

        found <- TRUE
        break
      }
    }
    if (found) break
  }
  expect_true(found, info = "No non-null winset entry found in payload.")
})

test_that("winset_num_samples affects boundary vertex count", {
  skip_without_lib()
  trace <- .ws_trace_2d()
  w16 <- animate_competition_canvas(
    trace, voters = WS_VOTERS, compute_winset = TRUE,
    winset_num_samples = 16L
  )
  w64 <- animate_competition_canvas(
    trace, voters = WS_VOTERS, compute_winset = TRUE,
    winset_num_samples = 64L
  )

  # Find the first non-null entry in each and compare path vertex counts.
  .first_path_len <- function(frame_payload) {
    for (f in seq_along(frame_payload)) {
      for (s_i in seq_along(frame_payload[[f]]$winsets)) {
        entry <- frame_payload[[f]]$winsets[[s_i]]
        if (!is.null(entry) && length(entry$polygons) > 0L) {
          return(length(entry$polygons[[1L]]$ring))
        }
      }
    }
    NA_integer_
  }

  len16 <- .first_path_len(w16$x$overlays_frames)
  len64 <- .first_path_len(w64$x$overlays_frames)
  expect_false(is.na(len16), info = "No non-null winset path found at num_samples=16.")
  expect_false(is.na(len64), info = "No non-null winset path found at num_samples=64.")
  # Higher num_samples should produce more vertices (or at least not fewer).
  expect_gte(len64, len16)
})

File: python/tests/test_competition_canvas_winset.py
# test_competition_canvas_winset.py — Unit tests for animate_competition_canvas WinSet feature.

import json
import re

import numpy as np
import pytest

import socialchoicelab as scl
import socialchoicelab.plots as sclp

# ---------------------------------------------------------------------------
# Fixture helpers
# ---------------------------------------------------------------------------

# 3 2D voters, 2 sticker competitors, 1 seat, 2 rounds.
# Competitor 0 at (0,0) wins every round (closer to 2 of 3 voters).
# n_frames = 3 (Round 1, Round 2, Final), 1 seat holder per frame.

WS_VOTERS = np.array([[-0.5, 0.0], [0.0, 0.2], [1.8, 0.1]])
_WS_COMPETITORS = np.array([[0.0, 0.0], [2.0, 0.0]])
_WS_ENGINE_CFG = scl.CompetitionEngineConfig(
    seat_count=1,
    seat_rule="plurality_top_k",
    max_rounds=2,
    step_config=scl.CompetitionStepConfig(kind="fixed", fixed_step_size=0.5),
)


def _ws_trace_2d():
    return scl.competition_run(
        _WS_COMPETITORS,
        ["sticker", "sticker"],
        WS_VOTERS,
        dist_config=scl.DistanceConfig(salience_weights=[1.0, 1.0]),
        engine_config=_WS_ENGINE_CFG,
    )


def _extract_payload(html: str) -> dict:
    """Parse the JSON payload embedded in the self-contained HTML widget."""
    match = re.search(
        r"HTMLWidgets\._init\(.*?\d+,\s*\d+,\s*(\{.*\})\s*\);",
        html,
        re.DOTALL,
    )
    assert match is not None, "Could not find JSON payload in HTML."
    return json.loads(match.group(1))


# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------


def test_compute_winset_false_leaves_key_absent():
    with _ws_trace_2d() as trace:
        html = sclp.animate_competition_canvas(trace, voters=WS_VOTERS)
    payload = _extract_payload(html)
    assert "winsets" not in payload
    assert "winset_competitor_indices" not in payload


def test_compute_winset_true_produces_correctly_shaped_payload():
    with _ws_trace_2d() as trace:
        n_rounds, n_competitors, _ = trace.dims()
        html = sclp.animate_competition_canvas(
            trace,
            voters=WS_VOTERS,
            compute_winset=True,
        )
    payload = _extract_payload(html)

    assert "winsets" in payload
    assert "winset_competitor_indices" in payload

    n_frames = n_rounds + 1  # 3
    ws = payload["winsets"]
    ws_idx = payload["winset_competitor_indices"]

    assert len(ws) == n_frames
    assert len(ws_idx) == n_frames

    for f in range(n_frames):
        frame_ws = ws[f]
        frame_idx = ws_idx[f]
        assert len(frame_ws) >= 1
        assert len(frame_ws) == len(frame_idx)


def test_compute_winset_without_voters_raises():
    with _ws_trace_2d() as trace:
        with pytest.raises(ValueError, match="voters"):
            sclp.animate_competition_canvas(trace, compute_winset=True)


def test_compute_winset_max_exceeded_raises():
    with _ws_trace_2d() as trace:
        # 1 seat × 3 frames = 3 > winset_max = 1.
        with pytest.raises(ValueError, match="winset_max"):
            sclp.animate_competition_canvas(
                trace,
                voters=WS_VOTERS,
                compute_winset=True,
                winset_max=1,
            )


def test_non_empty_winset_entry_has_required_keys():
    with _ws_trace_2d() as trace:
        n_rounds, n_competitors, _ = trace.dims()
        html = sclp.animate_competition_canvas(
            trace,
            voters=WS_VOTERS,
            compute_winset=True,
        )
    payload = _extract_payload(html)
    ws = payload["winsets"]

    # Find the first non-null entry.
    found = False
    for frame_ws in ws:
        for entry in frame_ws:
            if entry is not None:
                assert "paths" in entry
                assert "is_hole" in entry
                assert "competitor_idx" in entry

                # paths is a non-empty list; each path is flat [x,y,...].
                assert len(entry["paths"]) >= 1
                first_path = entry["paths"][0]
                assert isinstance(first_path, list)
                assert len(first_path) >= 4
                assert all(isinstance(v, float) for v in first_path)

                # is_hole length matches paths length.
                assert len(entry["is_hole"]) == len(entry["paths"])

                # competitor_idx is 0-based.
                assert 0 <= entry["competitor_idx"] < n_competitors

                found = True
                break
        if found:
            break

    assert found, "No non-null winset entry found in payload."


def test_winset_num_samples_affects_boundary_vertex_count():
    with _ws_trace_2d() as trace:
        html16 = sclp.animate_competition_canvas(
            trace, voters=WS_VOTERS, compute_winset=True, winset_num_samples=16
        )
        html64 = sclp.animate_competition_canvas(
            trace, voters=WS_VOTERS, compute_winset=True, winset_num_samples=64
        )

    def _first_path_len(ws_payload):
        for frame_ws in ws_payload:
            for entry in frame_ws:
                if entry is not None and entry["paths"]:
                    return len(entry["paths"][0])
        return None

    ws16 = _extract_payload(html16)["winsets"]
    ws64 = _extract_payload(html64)["winsets"]
    len16 = _first_path_len(ws16)
    len64 = _first_path_len(ws64)
    assert len16 is not None, "No non-null winset path found at num_samples=16."
    assert len64 is not None, "No non-null winset path found at num_samples=64."
    # Higher num_samples should produce more vertices (or at least not fewer).
    assert len64 >= len16
